crates/bevy_render/src/batching/mod.rs

```diff
@@ -0,0 +1,135 @@
+use bevy_ecs::{
+    component::Component,
+    prelude::Res,
+    query::{ReadOnlyWorldQuery, WorldQuery},
+    system::{Query, ResMut},
+};
+use nonmax::NonMaxU32;
+
+use crate::{
+    render_phase::{CachedRenderPipelinePhaseItem, DrawFunctionId, RenderPhase},
+    render_resource::{CachedRenderPipelineId, GpuArrayBuffer, GpuArrayBufferable},
+    renderer::{RenderDevice, RenderQueue},
+};
+
+/// Add this component to mesh entities to disable automatic batching
+#[derive(Component)]
+pub struct NoAutomaticBatching;
+
+/// Data necessary to be equal for two draw commands to be mergeable
+///
+/// This is based on the following assumptions:
+/// - Only entities with prepared assets (pipelines, materials, meshes) are
+///   queued to phases
+/// - View bindings are constant across a phase for a given draw function as
+///   phases are per-view
+/// - `prepare_and_batch_meshes` is the only system that performs this batching
+///   and has sole responsibility for preparing the per-object data. As such
+///   the mesh binding and dynamic offsets are assumed to only be variable as a
+///   result of the `prepare_and_batch_meshes` system, e.g. due to having to split
+///   data across separate uniform bindings within the same buffer due to the
+///   maximum uniform buffer binding size.
+struct BatchMeta<T: PartialEq> {
+    /// The pipeline id encompasses all pipeline configuration including vertex
+    /// buffers and layouts, shaders and their specializations, bind group
+    /// layouts, etc.
+    pub pipeline_id: CachedRenderPipelineId,
+    /// The draw function id defines the RenderCommands that are called to
+    /// set the pipeline and bindings, and make the draw command
+    pub draw_function_id: DrawFunctionId,
+    pub dynamic_offset: Option<NonMaxU32>,
+    pub user_data: T,
+}
+
+impl<T: PartialEq> PartialEq for BatchMeta<T> {
+    #[inline]
+    fn eq(&self, other: &BatchMeta<T>) -> bool {
+        self.pipeline_id == other.pipeline_id
+            && self.draw_function_id == other.draw_function_id
+            && self.dynamic_offset == other.dynamic_offset
+            && self.user_data == other.user_data
+    }
+}
+
+pub trait GetBatchData {
+    type Query: ReadOnlyWorldQuery;
+    type CompareData: PartialEq;
+    type BufferData: GpuArrayBufferable + Sync + Send + 'static;
+    fn get_batch_data(
+        batch_data: <Self::Query as WorldQuery>::Item<'_>,
+    ) -> (Self::CompareData, Self::BufferData);
+}
+
+/// Batch the items in a render phase. This means comparing metadata needed to draw each phase item
+/// and trying to combine the draws into a batch.
+pub fn batch_and_prepare_render_phase<I: CachedRenderPipelinePhaseItem, F: GetBatchData>(
+    gpu_array_buffer: ResMut<GpuArrayBuffer<F::BufferData>>,
+    mut views: Query<&mut RenderPhase<I>>,
+    query: Query<(Option<&NoAutomaticBatching>, F::Query)>,
```

Comment:
this could use `Has<NoAutomaticBatching>` (nit picking)

Reply:
Done.